File: modules/agent/src/models/DataProvenance.py
"""DataProvenance model for field-level tracking."""

from sqlalchemy import Column, Text, DateTime, BigInteger, Numeric, ForeignKey, func, CheckConstraint
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from models import Base


class DataProvenance(Base):
    """DataProvenance SQLAlchemy model."""

    __tablename__ = "Data_Provenance"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    entity_type = Column(Text, nullable=False)     # 'Organization', 'Individual'
    entity_id = Column(BigInteger, nullable=False)
    field_name = Column(Text, nullable=False)      # 'revenue_range_id', 'seniority_level', etc.
    source = Column(Text, nullable=False)          # 'clearbit', 'apollo', 'linkedin', 'agent', 'manual'
    source_url = Column(Text, nullable=True)
    confidence = Column(Numeric(3, 2), nullable=True)  # 0.00 to 1.00
    verified_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    verified_by = Column(BigInteger, ForeignKey("User.id", ondelete="SET NULL"), nullable=True)  # NULL = AI agent
    raw_value = Column(JSONB, nullable=True)       # Original value from source
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    created_by = Column(BigInteger, ForeignKey("User.id"), nullable=False)
    updated_by = Column(BigInteger, ForeignKey("User.id"), nullable=False)

    # Relationships
    verifier = relationship("User", foreign_keys=[verified_by])

    __table_args__ = (
        CheckConstraint('confidence >= 0 AND confidence <= 1', name='confidence_range'),
    )

File: modules/agent/internal/models/DataProvenance.go
package models

import (
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/datatypes"
)

type DataProvenance struct {
	ID         int64            `gorm:"primaryKey;autoIncrement" json:"id"`
	EntityType string           `gorm:"type:text;not null" json:"entity_type"`  // Organization, Individual
	EntityID   int64            `gorm:"not null" json:"entity_id"`
	FieldName  string           `gorm:"type:text;not null" json:"field_name"`   // revenue_range_id, seniority_level, etc.
	Source     string           `gorm:"type:text;not null" json:"source"`       // clearbit, apollo, linkedin, agent, manual
	SourceURL  *string          `gorm:"type:text" json:"source_url"`
	Confidence *decimal.Decimal `gorm:"type:numeric(3,2)" json:"confidence"`    // 0.00 to 1.00
	VerifiedAt time.Time        `gorm:"autoCreateTime" json:"verified_at"`
	VerifiedBy *int64           `json:"verified_by"`                            // NULL = AI agent
	RawValue   datatypes.JSON   `gorm:"type:jsonb" json:"raw_value"`
	CreatedAt  time.Time        `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt  time.Time        `gorm:"autoUpdateTime" json:"updated_at"`
}

func (DataProvenance) TableName() string {
	return "Data_Provenance"
}
